packages/eslint-plugin/tests/rules/consistent-type-imports.test.ts

```diff
@@ -1215,5 +1222,24 @@ const a: Default = '';
         },
       ],
     },
+    {
```

Comment:
what about this case?
```ts
import type { A } from 'foo';
import { B } from 'foo';
let foo: A;
let bar = B;
```

This should error.

Reply:
Thought about that one as well, but I think such a rule should not target explicitly split imports and only ignore those where type and value imports are mixed.
So this was actually intentional

Reply:
I think that this is the wrong approach. As it stands right now, this option makes the lint rule do the following:

- if all imports in a statement are used as a type, enforce the statement is type only
- else do nothing.

Lint rules are about emforcing consistency - and that "else do nothing" is the opposite of consistency.

It should always enforce that there is a single import statement.

If someone has a use-case where they want to split the import out, then a user can opt-out of the consistency explicitly using an eslint-disable comment.

Reply:
I can see this also being an option, however
1. I see this variant being very useful as well and in my case it is the preferred rule behavior for my case
2. Adding the validation for this reverse situation (two imports from the same file) is way more difficult to check for and would require a lot more changes to the code

You could describe the behavior as this linting rule as "Do not allow value imports containing only types" which I would still say is a consistent rule behavior.

Could you perhaps agree to adding this kind of option as it working right now and then thinking about a variant which would also check for the reverse later on?

Reply:
I don't think that allowing an inconsistent state is the correct approach.
The lint rule should be enforcing one way or another.

I don't see the value in just enforcing one thing - it really not much better than not using a lint rule at all.
I think that this rule should be strict and consistent as per its namesake.

Reply:
Ok I should probably explain the reasoning in detail so you see why I mean it it very useful:
The benefit of this approach is to get code which is that when only type imports are used, the linting rule would help the user prevent unnecessary import references for files which are not actually required. Removing these imports would allow eg. webpack to properly know that there is no code dependency between both files which helps with stuff like hot-reloading, build times and more.
This far I think you can agree with me why a linting rule for type imports is important - and this can of course be achievable with the current rule as well.

The problem with the current `type-imports` prefer option however is that it wants to split import statements which contain type and non-type imports.
To prevent splitting these imports and unnecessarily modify those each time an important class is actually used as a value when the code is changed, the `type-imports-mixed` would allow the user to keep those imports in a single line.

However, compared to what you propose of **always** requiring the user to merge all imports of the same file into the same import statement, this rule option would still allow having a separate import for types in the file.
This can be useful in various cases:
- Eg. code which was initially written for `type-imports` would not result in errors once switching to `type-imports-mixed`
- Some users might prefer to to split import statements into two separate statements, when they have many type and value imports to create separation

To sum it up, this rule option would provide optimal performance benefit (no unnecessary imports to files which are only used as types), while still allowing maximum freedom for the coder.
And this freedom for the coder is what I think should still be allowed as an option. For me, enabling the option `type-imports` right now is just not possible because of it's aggressive behavior and this rule option is what would fix this.

Reply:
PS: Also, if you intend to have it implemented the way you think it should work, feel free to write it, however I have to say that I think that variant is way more difficult to implement and enforce I think, because all import statements in the file have to be validated together which would require a completely different code that the one right now. In fact, it would probably make sense to write a completely separate rule/implementation for such a behavior.

Reply:
I 100% understand the reasoning behind the option itself - I do - and I'm all for an option which allows you to reduce duplication and allow value imports of types if there is a value import.

However I don't agree with the "freedom" argument, and I don't agree that it should allow type-only imports when there is a value import.
Lint rules aren't meant to give freedom - they are supposed to restrict freedom to enforce consistency.

If developers need to break out of the box - this is the entire reason that `eslint-disable` comments exist.

---

Is it going to be harder to implement? For sure.
But there's no immediate rush or deadline here - we can take the time to do this right.

I'm okay with not implementing a fixer for this part of the option - as long as the lint rule detects and reports on it, that's enough.
Someone else can implement a fixer down the line if they so desire.

----

If you feel like you don't want to work on that - that's 100% okay.
This PR now exists as a starting point for anyone that wants to work on it in the future.

Also worth noting that this is all permissively licenced OSS.
If you don't want to work on this to the standard I am asking - you're welcome to fork the rule and use it privately or publish it as a separate plugin.

Reply:
Well I guess then I will have to copy the rule implementation into my custom rule project.

Also: Don't you think just adding the variant you proposed as yet another option would be fine as well? Especially because the implementation changes for the one I proposed are soo small that it would provide more usability which just these small changes?
I mean: what does it hurt to have this option as well?

Reply:
more options = more code paths.
more code paths = more maintenance.
This project is maintained by volunteers.
Keeping the maintenance burden low to help control the workload is a necessity. 

Additionally more options means that it's harder for users to get setup as they have to read through more docs and understand more things.
People already struggle to setup linting for their codebase due to the sheer number of rules in the ecosystem alone. Adding more options on top of that increases the complexity and makes it harder.

Simpler is better.

Options (for the most part) live forever.
Removing them takes conscious effort, a breaking change, and clear documentation to migrate off it.
Which means they are *hard* to remove. Additionally this just adds back into the aforementioned maintenance overhead.

Reply:
> Simpler is better.

Well this just clearly speaks for the solution I proposed... however if you still do not want to include this option, there's nothing I can do I guess.
I just think it's really bad that I will have to copy the rule and all the utils used into a separate new project just for the sake of adding a really simple option.

Is there really no way you can change your mind?
Right now the only reason you mentioned was that you do not like the behavior of the rule option. However, isn't this something the users will have to decide for themselves?

Reply:
@bradzacher could you perhaps reconsider this PR?
I would really like to refrain from having to fork this rule and probably no one would implement this useful feature otherwise.

Reply:
As mentioned - I'm happy to accept an option if it enforces that you combine imports **_always_**.
There must not be any exceptions - if the user wants an exception for some reason that's what an `eslint-disable` comment is for.

With the current logic - the rule just works inconsistently; users will be confused about why it forces certain imports to be combined but ignores other imports. This leads to users thinking the rule is broken, and reduces trust in the linter.

If you do not wish to make that change in your PR, then I'm sorry - I cannot accept it.